src/storage.rs

```diff
@@ -334,6 +334,138 @@ impl<T: Default> UnprotectedStorage<T> for NullStorage<T> {
     unsafe fn remove(&mut self, _: Index) -> T { Default::default() }
 }
 
+/// Wrapper storage that stores modifications to components in a bitset.
+///
+/// **Note: Never use `.iter()` on a mutable component storage that uses this.**
+///
+/// #Example Usage:
+/// 
+/// ```rust,ignore
+///# extern crate specs;
+///# use specs::{Planner, World, RunArg, Component, TrackedStorage, VecStorage, System, Join};
+/// 
+/// pub struct Comp(u32);
+/// impl Component for Comp {
+///     // `TrackedStorage` acts as a wrapper around another storage.
+///     // You can put any store inside of here (e.g. HashMapStorage, VecStorage, etc.)
+///     type Storage = TrackedStorage<Comp, VecStorage<Comp>>;
+/// }
+/// 
+/// pub struct CompSystem;
+/// impl System<()> for CompSystem {
+///     fn run(&mut self, arg: RunArg, _: ()) {
+///         let (entities, mut comps) = arg.fetch(|w| {
+///             (w.entities(), w.write::<Comp>()) 
+///         });
+/// 
+///         // Iterates over all components like normal.
+///         for (entity, comp) in (&entities, &comps).iter() {
+///             // ...
+///         }
+/// 
+///         // **Never do this**
+///         // This will flag all components as modified regardless of whether the inner loop
+///         // did modify their data.
+///         for (entity, comp) in (&entities, &mut comps).iter() {
+///             // ...
+///         }
+/// 
+///         // Instead do something like:
+///         for (entity, comp) in (&entities, &comps.check()).iter() {
+///             if true { // check whether you should modify this component or not.
+///                 let mut comp = comps.get_mut(entity);
+///                 // ...
+///             }
+///         }
+/// 
+///         // To iterate over the flagged/modified components:
+///         for (entity, flagged_comp) in (&entities, (&comps).open().1).iter() {
+///             // ...
+///         }
+/// 
+///         // Clears the tracked storage every frame with this system.
+///         (&mut comps).open().1.clear();
+///     }
+/// }
+///# fn main() { }
+/// ```
+pub struct TrackedStorage<C: Component, T: UnprotectedStorage<C>> {
+    mask: BitSet,
+    storage: T,
+    phantom: PhantomData<C>,
+}
+
+impl<C: Component, T: UnprotectedStorage<C>> UnprotectedStorage<C> for TrackedStorage<C, T> {
+    fn new() -> Self {
+        TrackedStorage {
+            mask: BitSet::new(),
+            storage: T::new(),
+            phantom: PhantomData,
+        }
+    }
+    unsafe fn clean<F>(&mut self, has: F) where F: Fn(Index) -> bool {
+        self.mask.clear();
+        self.storage.clean(has);
+    }
+    unsafe fn get(&self, id: Index) -> &C {
+        self.storage.get(id)
+    }
+    unsafe fn get_mut(&mut self, id: Index) -> &mut C {
+        // calling `.iter()` on a mutable reference to the storage will flag everything
+        self.mask.add(id);
```

Comment:
I imagine that such a `TrackedStorage` would be best used in the following way:

```rust
// pos is a component storage with a `TrackedStorage`
for (vel, pos) in (&vel, &mut pos).join() {
    *pos += vel * delta_time; 
}
```

In another system:

```rust
for (pos_handle, matrix) in (pos.check(), &mut mvp).join() {
    if pos.flagged(pos_handle) {
        // Recalc
    } else {
        // Use cached version
    }
}
```

It's inconvenient to do whatever checks and often too complicated, instead we might want a method on `UnprotectedStorage` which is called after every iterated component, where we check for equality. This would also solve the lifetime problem of a guard.